spec.emu

```diff
@@ -835,12 +1710,264 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-abstract-operations-for-modulesourceprototype-objects">
+      <h1>Abstract Operations for ModuleSourcePrototype Objects</h1>
+      <emu-clause id="sec-retrievehostsourcedata" type="abstract operation">
+        <h1>
+          RetrieveHostSourceData (
+            _anyModuleSourceInstance_: an Object,
+          ): anything
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>Used to retrieve the host-defined source data from a subclass of %ModuleSourcePrototype% to populate for a new instance of its module record by the host.</dd>
+        </dl>
+        <emu-alg>
+          1. Assert: _anyModuleSourceInstance_ has a [[HostSourceData]] internal slot.
+          1. Return _anyModuleSourceInstance_.[[HostSourceData]].
+        </emu-alg>
+      </emu-clause>
+
+      <emu-note>
+        <p>When creating a ModuleRecord for a given ModuleSource instance, this abstract method can be used by hosts to initialize the necessary [[HostDefined]] data on the ModuleRecord.</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-modulesource-constructor">
+      <h1>The ModuleSource Constructor</h1>
+      <p>The ModuleSource Object ModuleSourcePrototype subclass represents the source phase of SourceTextModuleRecord.</p>
+
+      <p>The ModuleSource constructor:</p>
+      <ul>
+        <li>is <dfn>%ModuleSource%</dfn>.</li>
+        <li>is the initial value of the *"ModuleSource"* property of the global object.</li>
+        <li>has a [[Prototype]] internal slot whose value is %ModuleSourcePrototype%.</li>
+        <li>does not have a [[HostSourceData]] internal slot.
+      </ul>
+    </emu-clause>
+
+    <emu-clause id="sec-properties-of-the-modulesource-prototype-object">
+      <h1>Properties of the ModuleSource Prototype Object</h1>
+      <p>The <dfn>ModuleSource prototype object</dfn>:</p>
+      <ul>
+        <li>is <dfn>%ModuleSource.prototype%</dfn>.</li>
+        <li>has a [[Prototype]] internal slot whose value is %ModuleSourcePrototype%.</li>
+        <li>is not a ModuleSource instance or an ModuleSourcePrototype instance and does not have a [[HostSourceData]] internal slot.
+        <li>has the following properties:</li>
+      </ul>
+
+      <emu-clause id="sec-get-modulesource-prototype-@@tostringtag">
+        <h1>get ModuleSource Prototype [ @@toStringTag ]</h1>
+        <p>`ModuleSource Prototype[@@toStringTag]` is an accessor property whose set accessor function is *undefined*. Its get accessor function performs the following steps when called:</p>
+        <emu-alg>
+          1. Let _O_ be ? ToObject(*this* value).
+          1. Perform ? RequireInternalSlot(_R_, [[HostSourceData]]).
+          1. Return the String value *"ModuleSource"*.
```

Comment:
Hardcoding the value in this method means that I can't use the base class method on every subclass to answer "which", only to answer "is". If, instead, we had a _second_ internal slot that contained "ModuleSource" by default, then subclasses would be encouraged to set the slot to their own value, and the base accessor would answer both "is" and "which" for all subclasses.

Reply:
This method is unique to the JS source. Other sources would need to hard-code their own values, as is mentioned explicitly in the spec note.

Note that ModuleSource is not supposed to be subclassed by other sources, unless it were a true derivation of JavaScript source text somehow.

Creating a special slot to store a string is an option as well. Although we would still have to just imply conformance by telling other source types that they should change this slot, so it doesn't seem that much of a win as conformance would still be by convention?

Reply:
Right - that's the point tho. I don't want to have to extract N accessors for N module types, I want one accessor to Rule Them All. In other words, this only really works if there exists a base class for all module types. If it's not ModuleSource, then we'd need AbstractModuleSource or similar.

Reply:
Is it satisfying if the base class has an accessor that responds based on an internal slot of the instance, such that the base class is not abstract but this behavior can be overridden? I don’t anticipate any JavaScript-specific properties on the module source base class, only properties with sensible semantics for any language of module.

Reply:
You can still construct such a generic function by reading the combination of the getter functions across the list of supported subclasses right?

We can support a generic `toStringTag` on `ModuleSourcePrototype`, the main thing would be a new internal slot to store the exact source type string, which I'm just not sure is enough to justify the change.

Reply:
The contents of the slot are expected to be overridden; ideally no module instance types would need to implement this accessor at all as a result - but if they did, then `baseAccessor.call(instance)` would continue to expose the slot value.

Reply:
I feel like in this case, we should use something more like a `get type () -> String` over a `Symbol.toStringTag` though.

Reply:
That's fine with me also.

Reply:
I like that `type` destructures and restructures with import attributes.
```js
const { type } = source;
await import(name, { with: { type } });
```
Or, at least it would if `type` were `undefined` for JavaScript, since `type` must be absent on the memo key for ECMAScript modules.

Reply:
`{ with: { ...(type && { type }) } }`

Reply:
Apropos of this, @erights attributes the aphorism “Things that are similar should be the same or different”, to Alan Kay but it doesn’t appear to be well-known. Regardless, maybe `ModuleSource.prototype.language` would be better since it’s similar but not the same as the `type` cache key.

Reply:
`sourceType`?

Reply:
Ok, I've implemented `sourceType` using a `ModuleSourceRecord` abstraction to avoid two separate slots. And with JS modules of type `sourceModule.sourceType === "source-text"`.

Reply:
> would be better since it’s similar but not the same as the `type` component of a module memo key.

I wrote about this in https://github.com/tc39/proposal-import-attributes/issues/140, pointing out that the `type` import attribute is still mostly an assertion. Ideally this could be `type` and the attribute would be renamed to `assertType`.